### DISCUSSION TITLE
How to find workflow with specific correlationId content optimally

### DISCUSSION THREAD

**Discussion by tsoG2000** [elsa-workflows/elsa-core#1621] [Q&A] [answered]
We were using _workflowInstanceStore.FindManyAsync() to find workflow with specific correlationId content. 
	Our correlationId consist of xml tags with GUIDs, e.g.: 

`<ParentCorrelationId>
			<Id>063f1203-174a-46b3-8011-393530e10b73</Id><workflowId>a3b1cdd6-a913-4287-9ddd-b4ae7f256365</workflowId> 
		</ParentCorrelationId>
		<ChildCorrelationId>
			<Id>063f1203-174a-46b3-8011-393530e10b73</Id><workflowId>a3b1cdd6-a913-4287-9ddd-b4ae7f256365</workflowId> 
		</ChildCorrelationId>`
		
	And we want to find are there any workflows already started (running or finished), with same Id. Looking into correlationid and tag <id></id> of all workflows. This is to prevent duplicate workflows for the same <id>.
	Is there better way of searching through correlationId's from Elsa for preventing duplicate workflow dispatch? Or some configuration that does it (preventing duplicates) automatically?

**Comment by sfmskywalker** [comment, marked as answer]
I think there are two parts to your question.

1. To efficiently query workflows by *a part of* the correlation ID  (since your correlation IDs are actually XML strings), you might try something like this:

```csharp

public class MyIdCorrelationSpecification : Specification<WorkflowInstance>
{
   MyIdCorrelationSpecification(string id) => CorrelationId = id;

   public string? CorrelationId { get; set; }

   public override Expression<Func<T, bool>> ToExpression() => x => x.CorrelationId.Contains(CorrelationId);
}

var specification = new MyIdCorrelationSpecification("some ID");
_workflowInstanceStore.FindMany(specification);
```

But that requires the persistence provider to support the `.Contains` method (EF Core does support this, so should be good).
Needless to say that this isn't as precise as actually parsing the XML string first and specifically looking for the `<Id>` element value. Some database servers do support querying XML values. If you wanted to take this route, you'd have to decorate the `IWorkflowInstanceStore` with a custom class that then explicitly handles the `MyIdCorrelationSpecification` to send a provider-specific query to the underlying provider.

That might work, but maybe not worth the trouble if doing a simple `Contains` suffices for your use case.

A better approach I think would be to associate the workflow instance with custom attributes that are then searchable. In your example, you would have three attributes: `ParentCorrelationId`, `ChildCorrelationId` and `Id`.

However, [that feature](https://github.com/elsa-workflows/elsa-core/issues/1524) doesn't exist yet.

2. Elsa does have a way to prevent duplicate workflow instance creation based on the correlation ID, but it does so by comparing the **whole** correlation ID value and does not offer a way to match against a part of the correlation ID. You would have to implement that query yourself (as outlined above).
